[export] ignore mark_dynamic() in export

pianpwk opened the pull request:
Previously we were accomodating `torch._dynamo.mark_dynamic()` for export's dynamic shapes. Here we clean things up and ignore it, requiring users to specify an export input for `dynamic_shapes`.

Note: there's 4 decorators relevant to export, `mark_dynamic, maybe_mark_dynamic, mark_static, mark_unbacked`. User calls that involve export have only been `mark_dynamic()`, and we use `maybe_mark_dynamic` under the hood for `Dim.AUTO`, but we could start using others. One reason I decided to not warn and just silently ignore is these decorators cause the tensors to carry dynamic info, and it'll be hard to tell whether the markers are from export or user calls when re-exporting with the same inputs.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @rec

avikchaudhuri commented on benchmarks/dynamo/common.py:
```diff
@@ -1355,6 +1356,16 @@ def load(cls, model, example_inputs, device):
+                # use this to produce dynamic_shapes spec instead.
+                if not isinstance(x, torch.Tensor):
+                    return None
+                return {
```
You sure you don't have to handle other markers?

pianpwk replied:
In these tests I'm pretty sure it's only mark_dynamic that's used..

avikchaudhuri commented on benchmarks/dynamo/common.py:
```diff
@@ -1364,10 +1375,14 @@ def load(cls, model, example_inputs, device):
             else:
                 _register_dataclass_output_as_pytree(example_outputs)
 
+            combined_args = tuple(example_args) + tuple(example_kwargs.values())
```
curious, no need to look at the model signature here?

pianpwk replied:
That could be better if the kwargs are out of order, but I don't think that's happening here. We'll just feed the shapes spec as a tuple.

avikchaudhuri commented on torch/export/dynamic_shapes.py:
```diff
@@ -846,16 +846,13 @@ def _tree_map_helper(tree, val):
             for i, val in enumerate(tensor.shape):
                 dim = shape.get(i, _DimHint.STATIC)
-                if _marked_dynamic(tensor, i) or dim == _DimHint.AUTO:
+                if dim == _DimHint.AUTO:
```
lol get rid of the dupe if

pianpwk committed:
[export] ignore mark_dynamic() in export (#135536)

Summary:
X-link: https://github.com/pytorch/benchmark/pull/2451

Previously we were accomodating `torch._dynamo.mark_dynamic()` for export's dynamic shapes. Here we clean things up and ignore it, requiring users to specify an export input for `dynamic_shapes`.

Note: there's 4 decorators relevant to export, `mark_dynamic, maybe_mark_dynamic, mark_static, mark_unbacked`. User calls that involve export have only been `mark_dynamic()`, and we use `maybe_mark_dynamic` under the hood for `Dim.AUTO`, but we could start using others. One reason I decided to not warn and just silently ignore is these decorators cause the tensors to carry dynamic info, and it'll be hard to tell whether the markers are from export or user calls when re-exporting with the same inputs.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec


Reviewed By: desertfire, avikchaudhuri

Differential Revision: D62404528

Pulled By: pianpwk
